polkadot/network/src/lib.rs

```diff
@@ -483,6 +509,64 @@ impl Specialization<Block> for PolkadotProtocol {
 
 	fn maintain_peers(&mut self, ctx: &mut Context<Block>) {
 		self.consensus_gossip.collect_garbage(None);
+		self.collators.collect_garbage(None);
 		self.dispatch_pending_requests(ctx);
+
+		for collator_action in self.collators.maintain_peers() {
+			match collator_action {
+				Action::Disconnect(collator) => self.disconnect_bad_collator(ctx, collator),
+				Action::NewRole(account_id, role) => if let Some(collator) = self.collator_peer_id(account_id) {
+					send_polkadot_message(
+						ctx,
+						collator,
+						Message::CollatorRole(role),
+					)
+				},
+			}
+		}
+	}
+
+	fn on_block_imported(&mut self, _ctx: &mut Context<Block>, hash: Hash, _header: &Header) {
+		self.collators.collect_garbage(Some(&hash));
+	}
+}
+
+impl PolkadotProtocol {
+	// we received a collation from a peer
+	fn on_collation(&mut self, ctx: &mut Context<Block>, from: PeerId, relay_parent: Hash, collation: Collation) {
+		let collation_para = collation.receipt.parachain_index;
+		let collated_acc = collation.receipt.collator;
+
+		match self.peers.get(&from) {
+			None => ctx.disconnect_peer(from),
+			Some(peer_info) => match peer_info.status.collating_for {
+				None => ctx.disable_peer(from),
+				Some((ref acc_id, ref para_id))
+					if para_id != &collation_para || acc_id != &collated_acc || collation.receipt.check_signature().is_err() => ctx.disable_peer(from),
+				Some((ref acc_id, _)) => self.collators.on_collation(acc_id.clone(), relay_parent, collation),
```

Comment:
`on_collation` above states that _we must have checked the signature_ prior, and although this happens in the match check before, arriving here, that hasn't necessarily happened, does it? Also relying on a check to have happened in the match prior sounds like something that could easily be missed when changing this later. I'd prefer to have a more explicit guard around it, that makes sure check_signature has been called before on_collation regardless. Something like 
```rust
Some((ref acc_id, ref para_id)) => {
  if para_id == &collation_para && acc_id == &collated_acc && collation.receipt.check_signature().is_ok() {
      self.collators.on_collation(acc_id.clone(), relay_parent, collation)
  } else {
    ctx.disable_peer(from)
  }
}
```

Reply:
have rewritten to this before seeing this comment. Agree